crates/driver/src/domain/competition/settled/mod.rs

```diff
@@ -0,0 +1,275 @@
+                    factor: _,
+                    max_volume_factor: _,
+                    quote: _,
+                } => Err(Error::UnsupportedFeePolicy),
+                order::FeePolicy::Volume { factor: _ } => Err(Error::UnsupportedFeePolicy),
+            }
+        };
+
+        let protocol_fee = self.policies.first().map(protocol_fee).transpose();
+        Ok(eth::Asset {
+            token: self.surplus_token(),
+            amount: protocol_fee?.unwrap_or(0.into()).into(),
+        })
+    }
+
+    /// Protocol fee is defined by fee policies attached to the order.
+    ///
+    /// Denominated in NATIVE token
+    fn native_protocol_fee(
+        &self,
+        prices: &auction::NormalizedPrices,
+    ) -> Result<eth::TokenAmount, Error> {
+        let protocol_fee = self.protocol_fee()?.amount;
+        let native_price = self.surplus_token_price(prices)?;
+        big_rational_to_u256(&(protocol_fee.0.to_big_rational() * native_price.0))
+            .map(Into::into)
+            .map_err(Into::into)
+    }
+
+    fn surplus_token(&self) -> eth::TokenAddress {
+        match self.side {
+            Side::Buy => self.sell.token,
+            Side::Sell => self.buy.token,
+        }
+    }
+
+    /// Returns the normalized price of the trade surplus token
+    fn surplus_token_price(
+        &self,
+        prices: &auction::NormalizedPrices,
+    ) -> Result<auction::NormalizedPrice, Error> {
+        prices
+            .get(&self.surplus_token())
+            .cloned()
+            .ok_or(Error::MissingPrice(self.surplus_token()))
+    }
+}
+
+fn apply_factor(amount: eth::U256, factor: f64) -> Option<eth::U256> {
+    Some(
+        amount.checked_mul(eth::U256::from_f64_lossy(factor * 1000000000000000000.))?
+            / 1000000000000000000u128,
+    )
+}
+
+/// Custom clearing prices at which the trade was executed.
+///
+/// These prices differ from uniform clearing prices, in that they are adjusted
+/// to account for all fees (gas cost and protocol fees).
+///
+/// These prices determine the actual traded amounts from the user perspective.
+#[derive(Debug, Clone)]
+pub struct CustomClearingPrices {
+    pub sell: eth::U256,
+    pub buy: eth::U256,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("multiple fee policies are not supported yet")]
+    MultipleFeePolicies,
+    #[error("failed to calculate surplus for trade sell {0:?} buy {1:?}")]
+    Surplus(eth::Asset, eth::Asset),
+    #[error("missing native price for token {0:?}")]
+    MissingPrice(eth::TokenAddress),
+    #[error("type conversion error")]
+    TypeConversion(#[from] anyhow::Error),
+    #[error("fee policy not supported")]
+    UnsupportedFeePolicy,
+    #[error("factor {1} multiplication with {0} failed")]
+    Factor(eth::U256, f64),
+    #[error("overflow error while calculating protocol fee")]
+    Overflow,
+    #[error("division by zero error while calculating protocol fee")]
+    DivisionByZero,
```

Comment:
I think we should avoid blindly adding a new error type for every potential call that could fail and instead ask us if we can group them in some sensible way. What is the advantage of having these strongly typed subtypes? I believe it's to then handle different types of errors differently (how are we realistically going to handle a `TypeConversion` differently from a `DivisionByZero`?)

Looking at `Factor` for instance, the only way this can happen is if the multiplication overflows (so really it's the same as Error::Overflow). I wonder if even the last three should be combined to something like `Math(anyhow::Error)`.

I think by and large there are three types of error here:

1. Math error (overflow, underflow, div by 0, etc)
2. Malformed auctions (ie. missing price, type conversions)
3. Limitations of the current implementation (multiple fee policies, unsupported policies)

Reply:
did some cleanup, not sure if good enough